### PR TITLE
Add union operators to MutableHeaders

### LINKED ISSUE #1239: Unable to merge MutableHeaders with dict
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Unable to merge `MutableHeaders` with `dict`.

### To reproduce

```python3
from starlette.datastructures import MutableHeaders

response = MutableHeaders()
response |= {"X-Header": "string"}
```
### Expected behavior

```python3
>>> response
MutableHeaders({'x-header': 'string'})
```

### Actual behavior

Raised exception:
```
TypeError: unsupported operand type(s) for |=: 'MutableHeaders' and 'dict'
```

### Environment

- OS: Ubuntu 21.04
- Python version: 3.9.6
- Starlette version: 0.15.0

### Additional context

Method `MutableHeaders.update()` can resolve the issue but I think comfortable is used to operators `|` (for `merge`) and `|=` (for `update`).

It's comfortable if you have many headers like:

```python3
...

headers = {
    "X-Header-1": "string-1",
    "X-Header-1": "string-2",
    "X-Header-3": "string-3",
    "X-Header-...": "string-...",
    "X-Header-N": "string-N",
}

response.headers |= headers  # equals to: response.headers.update(headers)

return response

...
```
It's like [Dictionary Merge & Update Operators in Python 3.9](https://docs.python.org/3/whatsnew/3.9.html#dictionary-merge-update-operators).

What do you think about it? Thank you.

### PR DESCRIPTION
**Pull request by manlix**
Fix for issue #1239 

Edit by @Kludex: Closes #1239

### CODE REVIEW

File: starlette/datastructures.py [hunk 1]
```diff
@@ -604,6 +604,21 @@ def __delitem__(self, key: str) -> None:
...
+            self[key] = val
+        return self
+
+    def __or__(self, other: typing.Union[dict, "MutableHeaders"]) -> "MutableHeaders":
+        new = self.mutablecopy()
+        if isinstance(other, MutableHeaders):
+            new.update(dict(other))
+        else:
+            new.update(other)
+        return new
```
Comment (jcwilson) [inline starlette/datastructures.py]: why not just implement them in terms of `update()` and forego the typechecking?

```
    def __ior__(self, other: typing.Union[dict, "MutableHeaders"]) -> "MutableHeaders":
        self.update(other)
        return self

    def __or__(self, other: typing.Union[dict, "MutableHeaders"]) -> "MutableHeaders":
        new = self.mutablecopy()
        new.update(other)
        return new
```

Comment (manlix) [reply starlette/datastructures.py]: @jcwilson Absolutely right. Changes have been applied. Thank you.

File: starlette/datastructures.py [hunk 2]
```diff
@@ -604,6 +604,15 @@ def __delitem__(self, key: str) -> None:
...
             del self._list[idx]
 
+    def __ior__(self, other: dict) -> "MutableHeaders":
+        self.update(other)
+        return self
+
+    def __or__(self, other: dict) -> "MutableHeaders":
+        new = self.mutablecopy()
+        new.update(other)
+        return new
```
Comment (Kludex) [inline starlette/datastructures.py]: I guess we'd want to make sure it's a mapping before performing the `__or__` operation.

```suggestion
    def __or__(self, other: object) -> "MutableHeaders":
        if not isinstance(other, Mapping):
            return NotImplemented
        new = self.mutablecopy()
        new.update(other)
        return new
```

**Review by Kludex — DISMISSED** [review]

**Comment by Kludex** [comment]
@manlix Are you still interested in this PR? If so, would you mind rebasing and checking my comment?

**Comment by manlix** [comment]
@Kludex Thank you for your suggest. PR has been updated + new tests bundle.

File: tests/test_datastructures.py [hunk 3]
```diff
@@ -162,6 +162,50 @@ def test_mutable_headers():
...
+    assert isinstance(h, MutableHeaders)
+    assert dict(h) == {"a": "1"}
+    assert h.items() == [("a", "1")]
+    assert h.raw == [(b"a", b"1")]
+
+
+def test_mutable_headers_merge_not_mapping():
+    h = MutableHeaders()
+    with pytest.raises(NotImplementedError):
+        h |= {"not_mapping"}
```
Comment (lovelydinosaur) [inline tests/test_datastructures.py]:
```suggestion
        h |= {"not_mapping"}  # type: ignore
```

File: tests/test_datastructures.py [hunk 4]
```diff
@@ -162,6 +162,50 @@ def test_mutable_headers():
...
+    with pytest.raises(NotImplementedError):
+        h | {"not_mapping"}
```
Comment (lovelydinosaur) [inline tests/test_datastructures.py]:
```suggestion
        h | {"not_mapping"}  # type: ignore
```

File: starlette/datastructures.py [hunk 5]
```diff
@@ -615,6 +615,19 @@ def __delitem__(self, key: str) -> None:
         for idx in reversed(pop_indexes):
             del self._list[idx]
 
+    def __ior__(self, other: typing.Mapping) -> "MutableHeaders":
+        if not isinstance(other, typing.Mapping):
+            raise NotImplementedError
```
Comment (lovelydinosaur) [inline starlette/datastructures.py]:
```suggestion
            raise TypeError(f"Expected a mapping but got {other.__class__.__name__})
```

File: starlette/datastructures.py [hunk 6]
```diff
@@ -615,6 +615,19 @@ def __delitem__(self, key: str) -> None:
...
+        self.update(other)
+        return self
+
+    def __or__(self, other: typing.Mapping) -> "MutableHeaders":
+        if not isinstance(other, typing.Mapping):
+            raise NotImplementedError
```
Comment (lovelydinosaur) [inline starlette/datastructures.py]:
```suggestion
            raise TypeError(f"Expected a mapping but got {other.__class__.__name__})
```

File: tests/test_datastructures.py [hunk 7]
```diff
@@ -162,6 +162,50 @@ def test_mutable_headers():
...
+    h |= {"a": "1"}
+    assert isinstance(h, MutableHeaders)
+    assert dict(h) == {"a": "1"}
+    assert h.items() == [("a", "1")]
+    assert h.raw == [(b"a", b"1")]
+
+
+def test_mutable_headers_merge_not_mapping():
+    h = MutableHeaders()
+    with pytest.raises(NotImplementedError):
```
Comment (lovelydinosaur) [inline tests/test_datastructures.py]:
```suggestion
    with pytest.raises(TypeError):
```

File: tests/test_datastructures.py [hunk 8]
```diff
@@ -162,6 +162,50 @@ def test_mutable_headers():
...
+        h |= {"not_mapping"}
+    with pytest.raises(NotImplementedError):
```
Comment (lovelydinosaur) [inline tests/test_datastructures.py]:
```suggestion
    with pytest.raises(TypeError):
```

File: starlette/datastructures.py [hunk 9]
```diff
@@ -615,6 +615,19 @@ def __delitem__(self, key: str) -> None:
         for idx in reversed(pop_indexes):
             del self._list[idx]
 
+    def __ior__(self, other: typing.Mapping) -> "MutableHeaders":
+        if not isinstance(other, typing.Mapping):
+            raise TypeError(f"Expected a mapping but got {other.__class__.__name__})
```
Comment (lovelydinosaur) [inline starlette/datastructures.py]:
```suggestion
            raise TypeError(f"Expected a mapping but got {other.__class__.__name__}")
```

File: starlette/datastructures.py [hunk 10]
```diff
@@ -615,6 +615,19 @@ def __delitem__(self, key: str) -> None:
...
+        self.update(other)
+        return self
+
+    def __or__(self, other: typing.Mapping) -> "MutableHeaders":
+        if not isinstance(other, typing.Mapping):
+            raise TypeError(f"Expected a mapping but got {other.__class__.__name__})
```
Comment (lovelydinosaur) [inline starlette/datastructures.py]:
```suggestion
            raise TypeError(f"Expected a mapping but got {other.__class__.__name__}")
```

**Review by lovelydinosaur — APPROVED** [review]
👍 - Great thanks!
Made some minor tweaks. (`TypeError`, not `NotImplemented`)

**Pull request merged**